### PR TITLE
[Cluster Launcher] Allow attach/exec on uninitialized head node

### LINKED ISSUE #17525: [autoscaler] setup command failure bricks cluster
### What is the problem?

*Ray version and other system information (Python version, TensorFlow version, OS):* master

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):

Whenever I fail to update my cluster, the cluster becomes "unfindable/unreachable" for Ray commands. For example, `ray attach` and `ray exec` become broken.

Rather, the expected behavior is that `ray attach` and `ray exec` should work even if the setup commands didn't work properly.

cc @DmitriGekhtman @ijrsvt 

```
  New status: update-failed
  !!!
  SSH command failed.
  !!!

  Failed to setup head node.
(horovod) ➜  horovod git:(main) ✗ ray submit richard.yaml  mini_shuffle_s3.py -- --address=auto
Converting legacy cluster config to a multi node type cluster config. Multi-node-type cluster configs are the recommended format for configuring Ray clusters. See the docs for more information:
https://docs.ray.io/en/master/cluster/config.html#full-configuration
Loaded cached provider configuration
If you experience issues with the cloud provider, try re-running the command with --no-config-cache.
Head node (i-0c371fe1903227c86) is in state update-failed.
Traceback (most recent call last):
  File "/Users/rliaw/miniconda3/envs/horovod/bin/ray", line 8, in <module>
    sys.exit(main())
  File "/Users/rliaw/miniconda3/envs/horovod/lib/python3.7/site-packages/ray/scripts/scripts.py", line 1959, in main
    return cli()
  File "/Users/rliaw/miniconda3/envs/horovod/lib/python3.7/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/rliaw/miniconda3/envs/horovod/lib/python3.7/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/rliaw/miniconda3/envs/horovod/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/rliaw/miniconda3/envs/horovod/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/rliaw/miniconda3/envs/horovod/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/rliaw/miniconda3/envs/horovod/lib/python3.7/site-packages/ray/scripts/scripts.py", line 1258, in submit
    down=False)
  File "/Users/rliaw/miniconda3/envs/horovod/lib/python3.7/site-packages/ray/autoscaler/_private/commands.py", line 1127, in rsync
    config, config_file, override_cluster_name, create_if_needed=False)
  File "/Users/rliaw/miniconda3/envs/horovod/lib/python3.7/site-packages/ray/autoscaler/_private/commands.py", line 1228, in _get_running_head_node
    config["cluster_name"]))
RuntimeError: Head node of cluster (shuffling-data-loader-horovod) not found!
(horovod) ➜  horovod git:(main) ✗
```

### PR DESCRIPTION
**Pull request by ijrsvt**
## Related issue number

Closes #17525

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(
   - [x] Manually tested

After a failed `setup_command`:
```
(ray) ubuntu@ip-MY_IP:~/ray$ ray attach myaws.yaml 
Loaded cached provider configuration
If you experience issues with the cloud provider, try re-running the command with --no-config-cache.
Head node (i-xxxx) is in state update-failed.
The head node being returned: i-xxxx is not `up-to-date`. If you are not debugging a startup issueit is recommended to restart this head node with:   ray stop  myaws.yaml
Fetched IP: PUBLIC_IP
(base) ray@ip-PRIVATE_IP:~$ 
```

```
(ray) ubuntu@ip-MY_IP:~/ray$ ray exec  myaws.yaml whoami
Loaded cached provider configuration
If you experience issues with the cloud provider, try re-running the command with --no-config-cache.
Head node (i-xxxx) is in state update-failed.
The head node being returned: i-xxxx is not `up-to-date`. If you are not debugging a startup issue it is recommended to restart this head node with:   ray stop  myaws.yaml
Fetched IP: PUBLIC_IP
Warning: Permanently added 'PUBLIC_IP' (ECDSA) to the list of known hosts.
ray
Shared connection to PUBLIC_IP closed.
```

### CODE REVIEW

File: python/ray/autoscaler/_private/commands.py [hunk 1]
```diff
@@ -1217,12 +1241,25 @@ def _get_running_head_node(config: Dict[str, Any],
...
-            create_if_needed=False)
+            create_if_needed=False,
+            _allow_uninitialized_state=False)
     else:
+        if _allow_uninitialized_state and _backup_head_node is not None:
+            cli_logger.warning(
+                f"The head node being returned: {_backup_head_node} is not "
+                "`up-to-date`. If you are not debugging a startup issue "
+                "it is recommended to restart this head node with: {}",
+                cf.bold(f"  ray stop  {printable_config_file}"))
```
Comment (richardliaw) [inline python/ray/autoscaler/_private/commands.py:1260]: Ray stop wouldn't work here; I think you mean `ray down` or `ray up`?

File: python/ray/autoscaler/_private/commands.py [hunk 2]
```diff
@@ -921,7 +921,8 @@ def exec_cluster(config_file: str,
                  override_cluster_name: Optional[str] = None,
                  no_config_cache: bool = False,
                  port_forward: Optional[Port_forward] = None,
-                 with_output: bool = False) -> str:
+                 with_output: bool = False,
+                 _allow_uninitialized_state: bool = False) -> str:
```
Comment (DmitriGekhtman) [inline python/ray/autoscaler/_private/commands.py:925]: Looks like `ray attach` defaults this flag to `True`. Why do we default to `False` for `ray exec`?

Comment (richardliaw) [reply python/ray/autoscaler/_private/commands.py:925]: Actually, why do we need to expose this? can this just always be True?

File: python/ray/autoscaler/_private/commands.py (hunk 1 above)
Comment (ijrsvt) [reply python/ray/autoscaler/_private/commands.py:1260]:
```suggestion
                cf.bold(f"  ray down  {printable_config_file}"))
```

Comment (ijrsvt) [reply python/ray/autoscaler/_private/commands.py:1260]: OOps! I meant `ray down`!

File: python/ray/autoscaler/_private/commands.py (hunk 2 above)
Comment (ijrsvt) [reply python/ray/autoscaler/_private/commands.py:925]: This is set to `True` only from the CLI call to `exec_cluster` in `scripts/scripts.py`. This is to ensure that other cases only work on a running cluster (specifically `ray submit`, `ray dashboard` and `ray monitor`).

**Review by richardliaw — APPROVED** [review]
works great for me :) thanks a bunch for the fast turnaround!

**Review by DmitriGekhtman — APPROVED** [review]

**Pull request merged**